keps/sig-auth/4317-pod-certificates/README.md

```diff
@@ -0,0 +1,1430 @@
+authenticating to kube-apiserver, with the same attributes provided by bound
+service account tokens:
+* A PodIdentity X.509 extension to allow an X.509 certificate to carry all
+  details of a pod's identity, including its UIDs and the node it is running on.
+* A new TLS user-mapping capability in kube-apiserver to understand client
+  certificates that embed PodIdentity extensions.
+* A new PodCertificateRequest signer
+  (`kubernetes.io/kube-apiserver-client-pod`), shipped in
+  kube-controller-manager, to issue compatible client certificate to pods.
+* An enhancement to client-go's InClusterConfig that automatically loads a
+  credential bundle from a well-known filesystem path and uses it as a client
+  certificate.
+
+While this first application motivates some of the design decisions in this KEP,
+it is explicitly intended that PodCertificateRequests and PodCertificate
+projected volume sources will be useful to third-party projects that want to
+easily and securely issue X.509 certificates to workloads running in a
+Kubernetes cluster.
+
+There is a draft implementation of the design outlined in this KEP:
+https://github.com/kubernetes/kubernetes/pull/127251
+
+## Motivation
+
+### Goals
+
+**Ensure that certificate issuance can be accomplished without relying on bearer
+tokens at any step**:  One of the most significant benefits of mTLS
+authentication is that the credential is split into two pieces (private key and
+certificate), only one of which is ever transmitted to the counterparty (and
+thus subject to eavesdropping or theft by the counterparty).  In order to
+preserve this guarantee, it's important that bearer tokens are not involved *at
+any step*, otherwise the security of the scheme devolves to just being a bearer
+token.  In this KEP, as long as both kubelet and the signer implementation
+authenticate to kube-apiserver using their own separately-bootstrapped client
+certificates, then no bearer tokens are involved in certificate issuance.
+
+**Make it easy to ship secure third-party signer implementations**:  Securely
+implementing and operating a third-party signer implementation today requires
+intimate familiarity with Kubernetes' security model, otherwise your signer may
+negate Kubernetes' node restriction security boundary.  This KEP moves
+responsibility for enforcing node restriction on PodCertificateRequests into
+kube-apiserver, so that signer implementations do not need to consider it.
+
+**Support issuance of both client and server certificates**: While this KEP does
+not include a concrete signer that issues dedicated server certificates,
+third-party signer implementations should be able to do so on top of the
+machinery described in this KEP.  Additionally, the certificates isssed by
+`kubernetes.io/kube-apiserver-client-pod` specify key and certificate usages
+that permit the certificates to be used for TLS servers, although they will not
+work with standard TLS server certificate validation.
```

Comment:
I don't feel strongly about this, but I included this for symmetry with tokens (which can be used for server identification).

My half-baked thought about why we should set the bits to allow server certificate usage:
* We have seen from Kubernetes service account JWTs that a rich ecosystem of usages for these JWTs (including both client and server authentication usages) has sprung up.
* Unlike tokens (which require a distinct audience for distinct purposes), certificates have no audience constraint (because they aren't bearer credentials, so they aren't vulnerable to the attacks that audience closes off).
* So the ecosystem that springs up around these certificates would likely collapse on using the default certificate (maybe by federating that CA that issued them into external identity systems).
* At least some of the users in that ecosystems would want to use the certificate as a server certificate.

Reply:
As documented above, we will omit the "server" usage.  Those who want to use the cert as a server cert will need to be writing custom TLS validation logic, and thus are free to ignore the lack of server usage bits.